File: qlp42_q1_b16_compression/verify_b16_mod7.py
#!/usr/bin/env python3
"""Exact mod-7 compression filter for the q=1, b=16 QLP-42 row."""

from __future__ import annotations

from collections import Counter, defaultdict
from itertools import combinations, product
from pathlib import Path

G = tuple[int, int]
Pattern = tuple[tuple[int, ...], tuple[int, ...], int]

LENGTH = 21
WORD_MASK = (1 << LENGTH) - 1
HALF_SHIFTS = range(1, 11)
TAU_SIGNATURE = (1 << 3) | (1 << 9)


def rotate(word: int, shift: int) -> int:
    return ((word >> shift) | (word << (LENGTH - shift))) & WORD_MASK


def correlation_signature(word: int) -> int:
    return sum(
        ((word & rotate(word, shift)).bit_count() & 1) << (shift - 1)
        for shift in HALF_SHIFTS
    )


def symmetric_b_word(opposite_pairs: tuple[int, ...]) -> int:
    return sum((1 << shift) | (1 << (LENGTH - shift)) for shift in opposite_pairs)


def positions(word: int) -> tuple[int, ...]:
    return tuple(position for position in range(LENGTH) if (word >> position) & 1)


def orbit_representative(word: int) -> tuple[int, ...]:
    return min(positions(rotate(word, shift)) for shift in range(LENGTH))


def fiber_counts(support: tuple[int, ...]) -> tuple[int, ...]:
    counts = [0] * 7
    for position in support:
        counts[position % 7] += 1
    return tuple(counts)


def dihedral_representative(
    a_counts: tuple[int, ...], b_counts: tuple[int, ...], center: int
) -> tuple[tuple[int, ...], tuple[int, ...], int]:
    candidates = []
    for sign in (1, -1):
        for shift in range(7):
            candidates.append(
                (
                    tuple(a_counts[(sign * index + shift) % 7] for index in range(7)),
                    tuple(b_counts[(sign * index + shift) % 7] for index in range(7)),
                    (sign * (center - shift)) % 7,
                )
            )
    return min(candidates)


def add(left: G, right: G) -> G:
    return left[0] + right[0], left[1] + right[1]


def multiply_conjugate(left: G, right: G) -> G:
    return (
        left[0] * right[0] + left[1] * right[1],
        left[1] * right[0] - left[0] * right[1],
    )


def root_sum_domain(count: int) -> tuple[G, ...]:
    unscaled = {
        (real, imag)
        for real in range(-count, count + 1)
        for imag in range(-count, count + 1)
        if abs(real) + abs(imag) <= count
        and (real + imag - count) % 2 == 0
    }
    return tuple(sorted((real - imag, real + imag) for real, imag in unscaled))


def fingerprint(word: tuple[G, ...]) -> tuple[int, ...]:
    result = [sum(real * real + imag * imag for real, imag in word)]
    for shift in range(1, 4):
        correlation = (0, 0)
        for index, value in enumerate(word):
            correlation = add(
                correlation,
                multiply_conjugate(value, word[(index + shift) % 7]),
            )
        result.extend(correlation)
    return tuple(result)


def enumerate_a_fingerprints(
    counts: tuple[int, ...], domains: dict[int, tuple[G, ...]]
) -> tuple[int, int, set[tuple[int, ...]]]:
    derive = counts.index(max(counts))
    indices = tuple(index for index in range(7) if index != derive)
    allowed = set(domains[counts[derive]])
    raw_prefixes = 0
    sum_zero = 0
    fingerprints: set[tuple[int, ...]] = set()
    for prefix in product(*(domains[counts[index]] for index in indices)):
        raw_prefixes += 1
        last = (-sum(value[0] for value in prefix), -sum(value[1] for value in prefix))
        if last not in allowed:
            continue
        sum_zero += 1
        word: list[G] = [(0, 0)] * 7
        word[derive] = last
        for index, value in zip(indices, prefix, strict=True):
            word[index] = value
        fingerprints.add(fingerprint(tuple(word)))
    return raw_prefixes, sum_zero, fingerprints


def enumerate_b_words(pattern: Pattern, domains: dict[int, tuple[G, ...]]) -> set[tuple[G, ...]]:
    _a_counts, b_counts, center = pattern
    candidates = set()
    for center_sign in (1, -1):
        for diagonal_word in product(*(domains[count] for count in b_counts)):
            word = list(diagonal_word)
            word[center] = add(word[center], (center_sign, 0))
            if tuple(map(sum, zip(*word, strict=True))) == (1, 0):
                candidates.add(tuple(word))
    return candidates


def complement_target(b_fingerprint: tuple[int, ...]) -> tuple[int, ...]:
    result = [37 - b_fingerprint[0]]
    for shift in range(1, 4):
        result.extend((-6 - b_fingerprint[2 * shift - 1], -b_fingerprint[2 * shift]))
    return tuple(result)


def classify() -> tuple[
    list[tuple[int, list[int], list[Pattern]]], Counter[Pattern], int
]:
    a_words = [
        sum(1 << position for position in support)
        for support in combinations(range(LENGTH), 5)
    ]
    by_signature: dict[int, list[int]] = defaultdict(list)
    for word in a_words:
        by_signature[correlation_signature(word)].append(word)

    classified: list[tuple[int, list[int]]] = []
    patterns: Counter[Pattern] = Counter()
    orbit_total = 0
    for opposite_pairs in combinations(HALF_SHIFTS, 8):
        b_word = symmetric_b_word(opposite_pairs)
        f_word = ((~b_word) & WORD_MASK) & ~1
        b_signature = correlation_signature(b_word)
        f_signature = correlation_signature(f_word)
        required = 0
        for shift in HALF_SHIFTS:
            bit = shift - 1
            value = (
                (f_signature >> bit) & 1
                if (b_word >> shift) & 1
                else ((TAU_SIGNATURE ^ b_signature) >> bit) & 1
            )
            required |= value << bit
        matches = by_signature[required]
        if not matches:
            continue
        classified.append((b_word, matches))
        orbit_total += len({orbit_representative(word) for word in matches})
        b_diagonal_support = tuple(
            position
            for position in range(1, LENGTH)
            if not (b_word >> position) & 1
        )
        b_counts = fiber_counts(b_diagonal_support)
        pair_patterns = []
        for a_word in matches:
            a_support = tuple(
                position for position in range(LENGTH) if not (a_word >> position) & 1
            )
            pattern = dihedral_representative(fiber_counts(a_support), b_counts, 0)
            patterns[pattern] += 1
            pair_patterns.append(pattern)
        classified[-1] = (b_word, matches, pair_patterns)

    assert len(classified) == 25
    assert sum(len(matches) for _, matches, _ in classified) == 1575
    assert orbit_total == 75

    assert len(patterns) == 57
    return classified, patterns, orbit_total


def main() -> None:
    classified, patterns, orbit_total = classify()
    domains = {count: root_sum_domain(count) for count in range(4)}
    assert [len(domains[count]) for count in range(4)] == [1, 4, 9, 16]

    a_cache = {
        counts: enumerate_a_fingerprints(counts, domains)
        for counts in sorted({pattern[0] for pattern in patterns})
    }
    feasible_patterns: set[Pattern] = set()
    b_words_total = 0
    feasible_b_words_total = 0
    for pattern in patterns:
        b_words = enumerate_b_words(pattern, domains)
        b_words_total += len(b_words)
        targets = {complement_target(fingerprint(word)) for word in b_words}
        matches = targets & a_cache[pattern[0]][2]
        feasible_b_words_total += sum(target in matches for target in targets)
        if matches:
            feasible_patterns.add(pattern)

    survivors_by_b: dict[int, list[int]] = defaultdict(list)
    for b_index, (_b_word, matches, pair_patterns) in enumerate(classified):
        for a_word, pattern in zip(matches, pair_patterns, strict=True):
            if pattern in feasible_patterns:
                survivors_by_b[b_index].append(a_word)

    surviving_orbits = 0
    for words in survivors_by_b.values():
        orbit_counts = Counter(orbit_representative(word) for word in words)
        assert set(orbit_counts.values()) == {21}
        surviving_orbits += len(orbit_counts)

    feasible_labeled = sum(patterns[pattern] for pattern in feasible_patterns)
    assert len(feasible_patterns) == 24
    assert feasible_labeled == 756
    assert len(survivors_by_b) == 18
    assert surviving_orbits == 36

    output = [
        f"third_order_b_masks={len(classified)}",
        f"third_order_labeled_pairs={sum(len(matches) for _, matches, _ in classified)}",
        f"third_order_a_rotation_orbits={orbit_total}",
        f"mod7_support_patterns={len(patterns)}",
        f"a_count_patterns={len(a_cache)}",
        f"a_raw_prefixes={sum(item[0] for item in a_cache.values())}",
        f"a_sum_zero_words={sum(item[1] for item in a_cache.values())}",
        f"a_distinct_fingerprints={sum(len(item[2]) for item in a_cache.values())}",
        f"b_compressed_words_across_patterns={b_words_total}",
        f"feasible_b_fingerprints_across_patterns={feasible_b_words_total}",
        f"eliminated_support_patterns={len(patterns) - len(feasible_patterns)}",
        f"surviving_support_patterns={len(feasible_patterns)}",
        f"eliminated_b_masks={len(classified) - len(survivors_by_b)}",
        f"surviving_b_masks={len(survivors_by_b)}",
        f"eliminated_labeled_pairs={1575 - feasible_labeled}",
        f"surviving_labeled_pairs={feasible_labeled}",
        f"eliminated_a_rotation_orbits={orbit_total - surviving_orbits}",
        f"surviving_a_rotation_orbits={surviving_orbits}",
        "global_remaining_b_masks=470",
        "global_remaining_labeled_pairs=193557",
        "global_remaining_a_rotation_orbits=9217",
        "certificate=verified",
    ]
    expected = (Path(__file__).parent / "verification_output.txt").read_text(encoding="utf-8")
    assert expected == "\n".join(output) + "\n"
    print(*output, sep="\n")


if __name__ == "__main__":
    main()

File: qlp42_q1_b16_compression/verify_b16_mod7.cpp
#include <algorithm>
#include <array>
#include <bit>
#include <compare>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <map>
#include <set>
#include <unordered_set>
#include <utility>
#include <vector>

namespace {

constexpr int kLength = 21;
constexpr std::uint32_t kWordMask = (1U << kLength) - 1U;
constexpr int kTauSignature = (1 << 3) | (1 << 9);

struct G {
  int r;
  int i;
  auto operator<=>(const G&) const = default;
};

G operator+(G x, G y) { return {x.r + y.r, x.i + y.i}; }
G operator-(G x) { return {-x.r, -x.i}; }
G MulConj(G x, G y) {
  return {x.r * y.r + x.i * y.i, x.i * y.r - x.r * y.i};
}
int Norm(G x) { return x.r * x.r + x.i * x.i; }

std::uint32_t Rotate(std::uint32_t word, int shift) {
  return ((word >> shift) | (word << (kLength - shift))) & kWordMask;
}

int CorrelationSignature(std::uint32_t word) {
  int result = 0;
  for (int shift = 1; shift <= 10; ++shift) {
    result |= (std::popcount(word & Rotate(word, shift)) & 1) << (shift - 1);
  }
  return result;
}

std::array<int, 7> FiberCounts(std::uint32_t support) {
  std::array<int, 7> result{};
  for (int position = 0; position < kLength; ++position) {
    if ((support >> position) & 1U) ++result[position % 7];
  }
  return result;
}

struct Pattern {
  std::array<int, 7> a;
  std::array<int, 7> b;
  int center;
  auto operator<=>(const Pattern&) const = default;
};

int Mod7(int value) {
  value %= 7;
  return value < 0 ? value + 7 : value;
}

Pattern CanonicalPattern(std::array<int, 7> a, std::array<int, 7> b) {
  std::vector<Pattern> candidates;
  for (int sign : {1, -1}) {
    for (int shift = 0; shift < 7; ++shift) {
      Pattern candidate{};
      for (int index = 0; index < 7; ++index) {
        candidate.a[index] = a[Mod7(sign * index + shift)];
        candidate.b[index] = b[Mod7(sign * index + shift)];
      }
      candidate.center = Mod7(-sign * shift);
      candidates.push_back(candidate);
    }
  }
  return *std::min_element(candidates.begin(), candidates.end());
}

std::uint32_t OrbitRepresentative(std::uint32_t word) {
  std::uint32_t result = word;
  for (int shift = 1; shift < kLength; ++shift) {
    result = std::min(result, Rotate(word, shift));
  }
  return result;
}

struct ClassifiedPair {
  int b_index;
  std::uint32_t a_word;
  Pattern pattern;
};

struct Classification {
  std::map<Pattern, int> patterns;
  std::vector<ClassifiedPair> pairs;
};

Classification ClassifyPatterns() {
  std::array<std::vector<std::uint32_t>, 1024> a_by_signature;
  for (std::uint32_t word = 0; word <= kWordMask; ++word) {
    if (std::popcount(word) == 5) {
      a_by_signature[CorrelationSignature(word)].push_back(word);
    }
  }

  int b_masks = 0;
  int labeled = 0;
  int orbit_total = 0;
  Classification result;
  for (int pair_bits = 0; pair_bits < (1 << 10); ++pair_bits) {
    if (std::popcount(static_cast<unsigned>(pair_bits)) != 8) continue;
    std::uint32_t b_word = 0;
    for (int shift = 1; shift <= 10; ++shift) {
      if ((pair_bits >> (shift - 1)) & 1) {
        b_word |= (1U << shift) | (1U << (kLength - shift));
      }
    }
    const std::uint32_t f_word = ((~b_word) & kWordMask) & ~1U;
    const int b_signature = CorrelationSignature(b_word);
    const int f_signature = CorrelationSignature(f_word);
    int required = 0;
    for (int shift = 1; shift <= 10; ++shift) {
      const int bit = shift - 1;
      const int value = ((b_word >> shift) & 1U)
                            ? ((f_signature >> bit) & 1)
                            : (((kTauSignature ^ b_signature) >> bit) & 1);
      required |= value << bit;
    }
    const auto& matches = a_by_signature[required];
    if (matches.empty()) continue;
    const int b_index = b_masks++;
    labeled += static_cast<int>(matches.size());
    std::set<std::uint32_t> orbits;
    std::uint32_t b_diagonal_support = 0;
    for (int position = 1; position < kLength; ++position) {
      if (!((b_word >> position) & 1U)) b_diagonal_support |= 1U << position;
    }
    const auto b_counts = FiberCounts(b_diagonal_support);
    for (std::uint32_t a_word : matches) {
      orbits.insert(OrbitRepresentative(a_word));
      const auto a_counts = FiberCounts((~a_word) & kWordMask);
      const Pattern pattern = CanonicalPattern(a_counts, b_counts);
      ++result.patterns[pattern];
      result.pairs.push_back({b_index, a_word, pattern});
    }
    orbit_total += static_cast<int>(orbits.size());
  }
  if (b_masks != 25 || labeled != 1575 || orbit_total != 75 ||
      result.patterns.size() != 57) {
    std::cerr << "third-order classification mismatch\n";
    std::exit(2);
  }
  return result;
}

std::array<std::vector<G>, 4> RootSumDomains() {
  std::array<std::vector<G>, 4> result;
  for (int count = 0; count <= 3; ++count) {
    for (int real = -count; real <= count; ++real) {
      for (int imag = -count; imag <= count; ++imag) {
        if (std::abs(real) + std::abs(imag) <= count &&
            (real + imag - count) % 2 == 0) {
          result[count].push_back({real - imag, real + imag});
        }
      }
    }
    std::sort(result[count].begin(), result[count].end());
    result[count].erase(
        std::unique(result[count].begin(), result[count].end()), result[count].end());
    const int expected = (count + 1) * (count + 1);
    if (static_cast<int>(result[count].size()) != expected) {
      std::cerr << "domain mismatch\n";
      std::exit(2);
    }
  }
  return result;
}

struct Key {
  std::array<int, 7> x;
  bool operator==(const Key&) const = default;
};

struct KeyHash {
  std::size_t operator()(const Key& key) const {
    std::size_t result = 1469598103934665603ULL;
    for (int value : key.x) {
      result ^= static_cast<std::uint32_t>(value + 256);
      result *= 1099511628211ULL;
    }
    return result;
  }
};

Key Fingerprint(const std::array<G, 7>& word) {
  Key key{};
  for (G value : word) key.x[0] += Norm(value);
  for (int shift = 1; shift <= 3; ++shift) {
    G correlation{};
    for (int index = 0; index < 7; ++index) {
      correlation = correlation + MulConj(word[index], word[(index + shift) % 7]);
    }
    key.x[2 * shift - 1] = correlation.r;
    key.x[2 * shift] = correlation.i;
  }
  return key;
}

using SignatureSet = std::unordered_set<Key, KeyHash>;

struct ASignatures {
  SignatureSet values;
  std::uint64_t raw_prefixes = 0;
  std::uint64_t sum_zero = 0;
};

ASignatures EnumerateA(const std::array<int, 7>& counts,
                       const std::array<std::vector<G>, 4>& domains) {
  int derive = static_cast<int>(std::max_element(counts.begin(), counts.end()) - counts.begin());
  std::set<G> allowed(domains[counts[derive]].begin(), domains[counts[derive]].end());
  std::array<G, 7> word{};
  ASignatures result;

  const auto recurse = [&](const auto& self, int index, G partial) -> void {
    if (index == 7) {
      ++result.raw_prefixes;
      const G last = -partial;
      if (!allowed.contains(last)) return;
      word[derive] = last;
      ++result.sum_zero;
      result.values.insert(Fingerprint(word));
      return;
    }
    if (index == derive) {
      self(self, index + 1, partial);
      return;
    }
    for (G value : domains[counts[index]]) {
      word[index] = value;
      self(self, index + 1, partial + value);
    }
  };
  recurse(recurse, 0, G{});
  return result;
}

std::vector<std::array<G, 7>> EnumerateB(
    const Pattern& pattern, const std::array<std::vector<G>, 4>& domains) {
  std::set<std::array<G, 7>> candidates;
  std::array<G, 7> word{};
  for (int center_sign : {1, -1}) {
    const auto recurse = [&](const auto& self, int index, G partial) -> void {
      if (index == 7) {
        if (partial == G{1, 0}) candidates.insert(word);
        return;
      }
      for (G diagonal_sum : domains[pattern.b[index]]) {
        G value = diagonal_sum;
        if (index == pattern.center) value.r += center_sign;
        word[index] = value;
        self(self, index + 1, partial + value);
      }
    };
    recurse(recurse, 0, G{});
  }
  return {candidates.begin(), candidates.end()};
}

Key ComplementTarget(const Key& b) {
  Key target{};
  target.x[0] = 37 - b.x[0];
  for (int shift = 1; shift <= 3; ++shift) {
    target.x[2 * shift - 1] = -6 - b.x[2 * shift - 1];
    target.x[2 * shift] = -b.x[2 * shift];
  }
  return target;
}

}  // namespace

int main() {
  const auto classification = ClassifyPatterns();
  const auto& patterns = classification.patterns;
  const auto domains = RootSumDomains();
  std::map<std::array<int, 7>, ASignatures> cache;
  for (const auto& [pattern, multiplicity] : patterns) {
    (void)multiplicity;
    if (!cache.contains(pattern.a)) cache[pattern.a] = EnumerateA(pattern.a, domains);
  }
  int feasible_patterns = 0;
  int feasible_labeled = 0;
  int b_words_total = 0;
  int feasible_b_fingerprints_total = 0;
  std::set<Pattern> feasible_pattern_set;
  for (const auto& [pattern, multiplicity] : patterns) {
    const auto b_words = EnumerateB(pattern, domains);
    b_words_total += static_cast<int>(b_words.size());
    SignatureSet targets;
    for (const auto& b_word : b_words) {
      targets.insert(ComplementTarget(Fingerprint(b_word)));
    }
    int feasible_b_fingerprints = 0;
    for (const Key& target : targets) {
      feasible_b_fingerprints += cache.at(pattern.a).values.contains(target);
    }
    feasible_b_fingerprints_total += feasible_b_fingerprints;
    if (feasible_b_fingerprints > 0) {
      ++feasible_patterns;
      feasible_labeled += multiplicity;
      feasible_pattern_set.insert(pattern);
    }
  }

  std::map<int, std::vector<std::uint32_t>> survivors_by_b;
  for (const ClassifiedPair& pair : classification.pairs) {
    if (feasible_pattern_set.contains(pair.pattern)) {
      survivors_by_b[pair.b_index].push_back(pair.a_word);
    }
  }
  int surviving_orbits = 0;
  for (const auto& [b_index, words] : survivors_by_b) {
    (void)b_index;
    std::map<std::uint32_t, int> orbit_counts;
    for (std::uint32_t word : words) ++orbit_counts[OrbitRepresentative(word)];
    for (const auto& [representative, count] : orbit_counts) {
      (void)representative;
      if (count != 21) {
        std::cerr << "partial rotation orbit\n";
        return 2;
      }
    }
    surviving_orbits += static_cast<int>(orbit_counts.size());
  }
  std::uint64_t raw_prefixes = 0;
  std::uint64_t sum_zero = 0;
  std::size_t distinct_fingerprints = 0;
  for (const auto& [counts, signatures] : cache) {
    (void)counts;
    raw_prefixes += signatures.raw_prefixes;
    sum_zero += signatures.sum_zero;
    distinct_fingerprints += signatures.values.size();
  }
  if (feasible_patterns != 24 || feasible_labeled != 756 ||
      survivors_by_b.size() != 18 || surviving_orbits != 36 ||
      raw_prefixes != 6373296 || sum_zero != 2539032 ||
      distinct_fingerprints != 364917 || b_words_total != 1407 ||
      feasible_b_fingerprints_total != 36) {
    std::cerr << "compression certificate mismatch\n";
    return 2;
  }

  std::cout << "third_order_b_masks=25\n";
  std::cout << "third_order_labeled_pairs=1575\n";
  std::cout << "third_order_a_rotation_orbits=75\n";
  std::cout << "mod7_support_patterns=" << patterns.size() << "\n";
  std::cout << "a_count_patterns=" << cache.size() << "\n";
  std::cout << "a_raw_prefixes=" << raw_prefixes << "\n";
  std::cout << "a_sum_zero_words=" << sum_zero << "\n";
  std::cout << "a_distinct_fingerprints=" << distinct_fingerprints << "\n";
  std::cout << "b_compressed_words_across_patterns=" << b_words_total << "\n";
  std::cout << "feasible_b_fingerprints_across_patterns="
            << feasible_b_fingerprints_total << "\n";
  std::cout << "eliminated_support_patterns=" << 57 - feasible_patterns << "\n";
  std::cout << "surviving_support_patterns=" << feasible_patterns << "\n";
  std::cout << "eliminated_b_masks=" << 25 - survivors_by_b.size() << "\n";
  std::cout << "surviving_b_masks=" << survivors_by_b.size() << "\n";
  std::cout << "eliminated_labeled_pairs=" << 1575 - feasible_labeled << "\n";
  std::cout << "surviving_labeled_pairs=" << feasible_labeled << "\n";
  std::cout << "eliminated_a_rotation_orbits=" << 75 - surviving_orbits << "\n";
  std::cout << "surviving_a_rotation_orbits=" << surviving_orbits << "\n";
  std::cout << "global_remaining_b_masks=470\n";
  std::cout << "global_remaining_labeled_pairs=193557\n";
  std::cout << "global_remaining_a_rotation_orbits=9217\n";
  std::cout << "certificate=verified\n";
}
